resources/type-system/inference.md

```diff
@@ -1324,51 +1360,100 @@ determined as follows:
 
   - Otherwise, let `K_1` be `FutureOr<K>`.
 
-- Let `m_1` be the result of performing expression type inference on `e_1`, in
+- Let `m_1` be the result of performing expression inference on `e_1`, in
   context `K_1`.
 
 - Let `T_1` be the static type of `m_1`.
 
-- Let `T` be `flatten(T_1)`.
-
-- Let `m` be an expression artifact whose runtime behavior is as follows:
-
-  - Execute expression artifact `m_1`, and let `o_1` be the resulting value. _By
-    expression soundness, `o_1` will be an instance of the type `T_1`._
-
-  - Define `o_2` as follows:
-
-    - If `o_1` is a subtype of `Future<T>`, then let `o_2` be `o_1`.
-
-    - Otherwise, let `o_2` be the result of creating a new object using the
-      constructor `Future<T>.value()` with `o_1` as its
-      argument.
-
-      - _TODO(paulberry): How do we ensure that the value passed to
-        `Future<T>.value()` is an instance of type `T`? What we have is that
-        `o_1` is an instance of `T_1`, where `T` is `flatten(T_1)`. Which is not
-        quite the same._
-
-    - _By expression soundness, `o_2` must be an instance of `Future<T>`._
-
-  - Pause the stream associated with the innermost enclosing asynchronous
-    **for** loop, if any.
-
-  - Suspend the current invocation of the function body immediately enclosing
-    the await expression until after `o_2` completes.
-
-  - At some time after `o_2` is completed, control returns to the current
-    invocation.
-
-  - If `o_2` is completed with an error `x` and stack trace `t`, then `m`
-    completes with the exception `x` and stack trace `t`. _This trivially
-    satisfies expression soundess, since expression soundness is only concerned
-    with the value of the expression when it completes normally._
-
-  - If `o_2` is completed with an object `v`, then `m` completes with the value
-    `v`. _Since `o_2` is an instance of `Future<T>`, by future soundness, `v`
-    must be an instance of `T`. This satisfies expression soundness, since `v`
-    is an instance of `T` , and `T` is the static type._
+- If `T_1` is incompatible with await (as defined in the _extension types_
+  specification), then there is a compile-time error.
+
+- Let `T_2` be `flatten(T_1)`.
+
+- Let `m_2` be `@LET(T_1 v = m_1 in v is Future<T_2> ? v :
+  Future<T_2>.value(@PROMOTED_TYPE<T_2>(v)))`.
+
+  - _Note that in many circumstances, it will be trivial for the compiler to
+    establish that `v is Future<T_2>` always evaluates to `true`, in which case
+    `m_2` can be optimized to `@PROMOTED_TYPE<Future<T_2>>(m_1)`._
+
+  - _For soundness, we must prove that whenever `@PROMOTED_TYPE<T_2>(v)`
```

Comment:
We should prove, once and for all, that `T <: FutureOr<flatten(T)>` for all types `T`.

Then here, on the `is! T_2` branch, with ` T_2 = flatten(T)`, you'd know that `T <: T_2`.

Reply:
Ok, I think I understand what you’re hinting at here. Fleshing out some details, you’re saying that if we know that `T <: FutureOr<flatten(T)>` for all `T`, we can substitute `T_1` for `T` and get that `T_1 <: FutureOr<flatten(T_1)>`, or equivalently, `T_1 <: FutureOr<T_2>`. In both branches, we know that `v` is an instance of `T_1`, therefore it must be an instance of `FutureOr<T_2>`. Clearly, then, there are two cases: `v` must either be an instance of `Future<T_2>` or an instance of `T_2`. The `is Future<T_2>` handles the first case, so obviously in the second case, `v` must be an instance of `T_2`.

I like it. It makes the motivation for this elaboration much clearer to me.

Since this suggestion requires separately proving that `T <: FutureOr<flatten(T)>` for all `T`, I’m going to postpone it to a separate PR.

Reply:
The current definition of *flatten* is ... not easy to work with (it's defined in terms of "derives a future type" which is a separate function, which means that the goals of *flatten* aren't obvious in "derives a future type", so it takes more proof to get from A to B. I've done it before, and I can probably do it again.

Reply:
This should be addressed by https://github.com/dart-lang/language/pull/3920.